tests/http/testenv/ports.py

```diff
@@ -24,24 +24,75 @@
 #
 ###########################################################################
 #
+import json
 import logging
+import os
+import re
 import socket
-from typing import Dict
+from json import JSONEncoder
+from typing import Dict, List
+
+import pytest
+from filelock import FileLock
 
 log = logging.getLogger(__name__)
 
 
-def alloc_ports(port_specs: Dict[str, int]) -> Dict[str, int]:
-    ports = {}
+def alloc_port_sets(n: int, port_specs: Dict[str, int]) -> List[Dict[str, int]]:
+    port_sets = []
     socks = []
-    for name, ptype in port_specs.items():
-        try:
-            s = socket.socket(type=ptype)
-            s.bind(('', 0))
-            ports[name] = s.getsockname()[1]
-            socks.append(s)
-        except Exception as e:
-            raise e
+    for _ in range(n):
+        ports = {}
+        for name, ptype in port_specs.items():
+            try:
+                s = socket.socket(type=ptype)
```

Comment:
It looks like there could be a race condition between the determination of free port numbers and the actual use later, if another port is opened in the meantime. Do the users of the port_sets handle that situation gracefully?

Reply:
Yes, that is a risk. I found no way around this though. The window of misfortune has always been there, but it increases with parallelism.

Reply:
I improved the port allocation strategy in the latest push. Now, when a server starts up, it just "allocates" the ports it needs with a retry logic should that fail. This is happening for every server started, shortening the time window between finding free ports and actually listening on them.